policy/options.go

```diff
@@ -0,0 +1,652 @@
+			if err != nil {
+				return err
+			}
+			normalizedURIDomains[i] = normalizedURIDomain
+		}
+		e.excludedURIDomains = normalizedURIDomains
+		return nil
+	}
+}
+
+func AddExcludedURIDomains(uriDomains []string) NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		normalizedURIDomains := make([]string, len(uriDomains))
+		for i, domain := range uriDomains {
+			normalizedURIDomain, err := normalizeAndValidateURIDomainConstraint(domain)
+			if err != nil {
+				return err
+			}
+			normalizedURIDomains[i] = normalizedURIDomain
+		}
+		e.excludedURIDomains = append(e.excludedURIDomains, normalizedURIDomains...)
+		return nil
+	}
+}
+
+func WithExcludedURIDomain(uriDomain string) NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		normalizedURIDomain, err := normalizeAndValidateURIDomainConstraint(uriDomain)
+		if err != nil {
+			return err
+		}
+		e.excludedURIDomains = []string{normalizedURIDomain}
+		return nil
+	}
+}
+
+func AddExcludedURIDomain(uriDomain string) NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		normalizedURIDomain, err := normalizeAndValidateURIDomainConstraint(uriDomain)
+		if err != nil {
+			return err
+		}
+		e.excludedURIDomains = append(e.excludedURIDomains, normalizedURIDomain)
+		return nil
+	}
+}
+
+func WithPermittedPrincipals(principals []string) NamePolicyOption {
+	return func(g *NamePolicyEngine) error {
+		// TODO(hs): normalize and parse principal into the right type? Seems the safe thing to do.
+		g.permittedPrincipals = principals
+		return nil
+	}
+}
+
+func WithExcludedPrincipals(principals []string) NamePolicyOption {
+	return func(g *NamePolicyEngine) error {
+		// TODO(hs): normalize and parse principal into the right type? Seems the safe thing to do.
+		g.excludedPrincipals = principals
+		return nil
+	}
+}
+
+func networkFor(ip net.IP) *net.IPNet {
+	var mask net.IPMask
+	if !isIPv4(ip) {
+		mask = net.CIDRMask(128, 128)
+	} else {
+		mask = net.CIDRMask(32, 32)
+	}
+	nw := &net.IPNet{
+		IP:   ip,
+		Mask: mask,
+	}
+	return nw
+}
+
+func isIPv4(ip net.IP) bool {
+	return ip.To4() != nil
+}
+
+func normalizeAndValidateDNSDomainConstraint(constraint string) (string, error) {
+	normalizedConstraint := strings.ToLower(strings.TrimSpace(constraint))
+	if strings.Contains(normalizedConstraint, "..") {
+		return "", errors.Errorf("domain constraint %q cannot have empty labels", constraint)
+	}
+	if strings.HasPrefix(normalizedConstraint, "*.") {
+		normalizedConstraint = normalizedConstraint[1:] // cut off wildcard character; keep the period
+	}
+	if strings.Contains(normalizedConstraint, "*") {
+		return "", errors.Errorf("domain constraint %q can only have wildcard as starting character", constraint)
+	}
+	if _, ok := domainToReverseLabels(normalizedConstraint); !ok {
+		return "", errors.Errorf("cannot parse permitted domain constraint %q", constraint)
+	}
+	return normalizedConstraint, nil
+}
```

Comment:
We might need to take into account international names too, see https://github.com/smallstep/crypto/blob/1648110a34d13e10c97e1feb5f059b3cfd23f3cb/tlsutil/utils.go#L11-L30

Reply:
And similar for any domain.